pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UnusedAssignmentRule.java

```diff
@@ -0,0 +1,1343 @@
+
+        Since we have the reaching definitions at each variable usage, we
+        could also use that to detect other kinds of bug, eg conditions
+        that are always true, or dereferences that will always NPE. In
+        the general case though, this is complicated and better left to
+        a DFA library, eg google Z3.
+
+        This analysis may be used as-is to detect switch labels that
+        fall-through, which could be useful to improve accuracy of other
+        rules.
+
+        TODO
+           * labels on arbitrary statements (currently only loops)
+           * explicit ctor call (hard to impossible without type res,
+             or at least proper graph algorithms like toposort)
+                -> this is pretty invisible as it causes false negatives, not FPs
+           * test ternary expr
+
+        DONE
+           * conditionals
+           * loops
+           * switch
+           * loop labels
+           * try/catch/finally
+           * lambdas
+           * constructors + initializers
+           * anon class
+           * test this.field in ctors
+           * foreach var should be reassigned from one iter to another
+           * test local class/anonymous class
+           * shortcut conditionals have their own control-flow
+           * parenthesized expressions
+           * conditional exprs in loops
+           * ignore variables that start with 'ignore'
+
+     */
+
+    private static final PropertyDescriptor<Boolean> CHECK_PREFIX_INCREMENT =
+        PropertyFactory.booleanProperty("checkUnusedPrefixIncrement")
+                       .desc("Report expressions like ++i that may be replaced with (i + 1)")
+                       .defaultValue(false)
+                       .build();
+
+    private static final PropertyDescriptor<Boolean> REPORT_UNUSED_VARS =
+        PropertyFactory.booleanProperty("reportUnusedVariables")
+                       .desc("Report variables that are only initialized, and never read at all. "
+                                 + "The rule UnusedVariable already cares for that, but you can enable it if needed")
+                       .defaultValue(false)
+                       .build();
+
+    public UnusedAssignmentRule() {
+        definePropertyDescriptor(CHECK_PREFIX_INCREMENT);
+        definePropertyDescriptor(REPORT_UNUSED_VARS);
+    }
+
+    @Override
+    public Object visit(ASTCompilationUnit node, Object data) {
+        for (JavaNode child : node.children()) {
+            if (child instanceof ASTTypeDeclaration) {
+
+                ASTAnyTypeDeclaration typeDecl = (ASTAnyTypeDeclaration) child.getChild(child.getNumChildren() - 1);
+                GlobalAlgoState result = new GlobalAlgoState();
+                typeDecl.jjtAccept(ReachingDefsVisitor.ONLY_LOCALS, new SpanInfo(result));
+
+                reportFinished(result, (RuleContext) data);
+            }
+        }
+
+        return data;
```

Comment:
I think, we can use rule chain, then? I'll try it before I merge....

Reply:
I've added the rule chain visit, but it probably doesn't make any difference anyway.